core/src/main/java/io/grpc/internal/InternalSubchannel.java

```diff
@@ -86,8 +86,19 @@
   /**
    * The index of the address corresponding to pendingTransport/activeTransport, or at beginning if
    * both are null.
+   *
+   * <p>Note: any {@link Index#updateAddresses(List)} should also update {@link #addressGroups}.
    */
-  private volatile Index addressIndex;
+  private Index addressIndex;
+
+  /**
+   * A volatile accessor to {@link Index#getAddressGroups()}. There are few methods ({@link
+   * #getAddressGroups()} and {@link #toString()} access this value where they supposed to access
+   * in the {@link #syncContext}. Ideally {@link Index#getAddressGroups()} can be volatile, so we
+   * don't need to maintain this volatile accessor. Although, having this accessor can reduce
```

Comment:
For posterity (it is fine to leave as-is): for me it was more of an argument for readability. With this approach the intent of the code is clearer.